cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -84,6 +84,373 @@ constexpr inline auto is_supported_construction_value_type()
+  static constexpr auto power30 = __uint128_t(power19) * 100000000000ULL;         ///< 10^30
+  static constexpr auto power31 = __uint128_t(power19) * 1000000000000ULL;        ///< 10^31
+  static constexpr auto power32 = __uint128_t(power19) * 10000000000000ULL;       ///< 10^32
+  static constexpr auto power33 = __uint128_t(power19) * 100000000000000ULL;      ///< 10^33
+  static constexpr auto power34 = __uint128_t(power19) * 1000000000000000ULL;     ///< 10^34
+  static constexpr auto power35 = __uint128_t(power19) * 10000000000000000ULL;    ///< 10^35
+  static constexpr auto power36 = __uint128_t(power19) * 100000000000000000ULL;   ///< 10^36
+  static constexpr auto power37 = __uint128_t(power19) * 1000000000000000000ULL;  ///< 10^37
+  static constexpr auto power38 = __uint128_t(power19) * power19;                 ///< 10^38
+};
+
+/**
+ * @brief Divide by a power of 10 that fits within a 32bit integer.
+ *
+ * @tparam T Type of value to be divided-from.
+ * @param value The number to be divided-from.
+ * @param exp10 The power-of-10 of the denominator, from 0 to 9 inclusive.
+ * @return Returns value / 10^exp10
+ */
+template <typename T>
+CUDF_HOST_DEVICE inline T divide_power10_32bit(T value, int exp10)
+{
+  // Computing division this way is much faster than the alternatives.
+  // Division is not implemented in GPU hardware, and the compiler will often implement it as a
+  // multiplication of the reciprocal of the denominator, requiring a conversion to floating point.
+  // Ths is especially slow for larger divides that have to use the FP64 pipeline, where threads
+  // bottleneck.
+
+  // Instead, if the compiler can see exactly what number it is dividing by, it can
+  // produce much more optimal assembly, doing bit shifting, multiplies by a constant, etc.
+  // For the compiler to see the value though, array lookup (with exp10 as the index)
+  // is not sufficient: We have to use a switch statement. Although this introduces a branch,
+  // it is still much faster than doing the divide any other way.
+  // Perhaps an array can be used in C++23 with the assume attribute?
+
+  // Since we're optimizing division this way, we have to do this for multiplication as well.
+  // That's because doing them in different ways (switch, array, runtime-computation, etc.)
+  // increases the register pressure on all kernels that use fixed_point types, specifically slowing
+  // down some of the PYMOD and join benchmarks.
+
+  // This is split up into separate functions for 32-, 64-, and 128-bit denominators.
+  // That way we limit the templated, inlined code generation to the exponents that are
+  // capable of being represented. Combining them together into a single function again
+  // introduces too much pressure on the kernels that use this code, slowing down their benchmarks.
+
+  // Also note that these divisors are hardcoded here in place, rather than in the PowersOf10
```

Comment:
I would guess that it's slower with the PowersOf10 struct because the members are `__uint128_t` whereas your literals below are probably all 32-bit ints.

I think if you template the return (and internal arithmetic) type of the recursive function template I gave above you should be able to use it here and get the same resulting performance / code.

Reply:
If I understand what you are suggesting correctly, this is what I tried with [this PR](https://github.com/rapidsai/cudf/pull/15110), which I had to roll back with [this PR](https://github.com/rapidsai/cudf/pull/15242).  It introduced too much performance degradation, for reasons I don't fully understand.

Reply:
I don't think so. In that PR you didn't have the switch. I'm just suggesting you can keep the switch, but get rid of the huge struct and the hard coded divisors by calling the function template at each of the switch cases.

Reply:
The gain is a reduction in code. I don't see why it wouldn't be computed at compile time even without `consteval` -- the recursion is on a non-type template parameter.

Reply:
I'll test the benchmarks and compilation time and see if there's any difference.

Reply:
There doesn't seem to be any difference, I'll change it to this